pkg/quarterdeck/middleware/auth.go

```diff
@@ -0,0 +1,218 @@
+	UserClaims                = "user_claims"
+	DefaultKeysURL            = "https://auth.ensign.app/.well-known/jwks.json"
+	DefaultAudience           = "https://ensign.app"
+	DefaultIssuer             = "https://auth.ensign.app"
+	DefaultMinRefreshInterval = 5 * time.Minute
+)
+
+// Authenticate middleware ensures that the request has a valid Bearer JWT in the
+// Authenticate header of the request otherwise it stops processing of the request and
+// returns a 401 unauthorized error. A valid Bearer JWT means that the access token is
+// supplied as the Bearer token, it is unexpired, and it was issued by Quarterdeck by
+// checking with the Quarterdeck public keys.
+//
+// In order to perform authorization, this middleware fetches public JSON Web Key Sets
+// (JWKS) from the authorizing Quarterdeck server and then caches them according to the
+// Cache-Control or Expires headers in the response. As Quarterdeck keys are rotated,
+// the cache must refresh the public keys in a background routine in order to correctly
+// authorize incoming JWT tokens. Users can control how the JWKS are fetched and cached
+// using AuthOptions (which are particularly helpful for tests).
+func Authenticate(opts ...AuthOption) (_ gin.HandlerFunc, err error) {
+	// Create the authorization options from the variadic arguments.
+	conf := NewAuthOptions(opts...)
+
+	// Create the JWK cache object using the context from the configuration
+	// This configuration tells the cache we want to refresh the JWKs when it needs to
+	// based on the Cache-Control or Expires header from he HTTP response. If the
+	// calculated minimum refresh interval is less than the configured minimum it won't
+	// refresh the JWKS any earlier. This means that the min refresh interval should be
+	// relatively small (e.g. minutes).
+	var validator tokens.Validator
+	if validator, err = conf.Validator(); err != nil {
+		return nil, err
+	}
+
+	return func(c *gin.Context) {
+		var (
+			err         error
+			accessToken string
+			claims      *tokens.Claims
+		)
+
+		// Get access token from the request
+		if accessToken, err = GetAccessToken(c); err != nil {
+			c.Error(err)
+			c.AbortWithStatusJSON(http.StatusUnauthorized, api.ErrorResponse(ErrAuthRequired))
+			return
+		}
+
+		// Verify the access token is authorized for use with Quarterdeck and extract claims.
+		if claims, err = validator.Verify(accessToken); err != nil {
+			c.Error(err)
+			c.AbortWithStatusJSON(http.StatusUnauthorized, api.ErrorResponse(ErrAuthRequired))
+			return
+		}
+
+		// Add claims to context for use in downstream processing and continue handlers
+		c.Set(UserClaims, claims)
+		c.Next()
+	}, nil
+}
+
+// GetAccessToken retrieves the bearer token from the authorization header and parses it
+// to return only the JWT access token component of the header. If the header is missing
+// or the token is not available, an error is returned.
+// TODO: switch to regular expression based parsing.
```

Comment:
Much faster processing for one, but also it can handle more cases such as multi-word tokens (which should not be allowed), case-insensitivity for "Bearer", different kinds of whitespace such as tabs or empty characters. It could even validate the token is a JWT token by checking that it is three base-64 encoded segments separated by ".".